pkg/cache/v3/linear.go

```diff
@@ -360,55 +484,48 @@ func (cache *LinearCache) UpdateResources(toUpdate map[string]types.Resource, to
 	defer cache.mu.Unlock()
 
 	cache.version++
-	cacheVersion := cache.getVersion()
-
-	modified := make(map[string]struct{}, len(toUpdate)+len(toDelete))
+	version := cache.getVersion()
+	modified := make([]string, 0, len(toUpdate)+len(toDelete))
 	for name, resource := range toUpdate {
-		cache.versionVector[name] = cacheVersion
-		cache.resources[name] = resource
-		modified[name] = struct{}{}
+		cache.resources[name] = newCachedResource(resource, version)
+		modified = append(modified, name)
 	}
 	for _, name := range toDelete {
-		delete(cache.versionVector, name)
 		delete(cache.resources, name)
-		modified[name] = struct{}{}
+		modified = append(modified, name)
 	}
 
-	cache.notifyAll(modified)
-
-	return nil
+	return cache.notifyAll(modified)
 }
 
 // SetResources replaces current resources with a new set of resources.
-// This function is useful for wildcard xDS subscriptions.
-// This way watches that are subscribed to all resources are triggered only once regardless of how many resources are changed.
+// If only some resources are to be updated, UpdateResources is more efficient.
 func (cache *LinearCache) SetResources(resources map[string]types.Resource) {
```

Comment:
Not something to necessarily change in this PR but it is weird that this specific method doesn't return an error.